python/tierkreis/controller/storage/walk.py

```diff
@@ -43,13 +41,14 @@ def walk_node(storage: ControllerStorage, node_location: NodeLocation) -> WalkRe
 
 
 def walk_eval(storage: ControllerStorage, node_location: NodeLocation) -> WalkResult:
+    logger.debug("walk_eval")
     walk_result = WalkResult([], [])
 
-    message = storage.read_output(node_location.append_node(0), Labels.THUNK)
-    graph = TierkreisGraph.from_proto(Graph.FromString(message))
+    message = storage.read_output(node_location.append_node(-1), Labels.THUNK)
+    graph = GraphData(**json.loads(message))
```

Comment:
I'm slightly surprised this works, does it actually work recursively? I wouldn't have expected `nodes` to actually be an instance of `NodeDef` with the default dataclasses library as I didn't think it did validation and construction of child classes.